modules/firewall-rules/variables.tf

```diff
@@ -0,0 +1,50 @@
+/**
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "project_id" {
+  description = "Project id of the project that holds the network."
+}
+
+variable "network_name" {
+  description = "Name of the network this set of firewall rules applies to."
+}
+
+variable "rules" {
+  description = "List of custom rule definitions (refer to variables file for syntax)."
+  default     = []
+  type = list(object({
```

Comment:
I see that we require `object` here but in the main module take in `any` and convert into this list of objects for this submodule? Is there a reason why we just take in `any` here and and do the defaults lookup here?

I am thinking if users want to use this submod outside of main module, they will need to define everything since its a list of objects.

Reply:
Ideally we should use list of objects everywhere. However as you pointed out users will need to define all variables possible through this. I chose to make the submodule list of objects to increase type checking while temporarily keeping the main module as type `any` until Terraform supports object defaults. 

I thought that was a decent compromise (if someone wants type checking they can use submodule but if they want a simpler interface then the main module implements that). Also, I was going to implement both as list of objects, but then the firewall rules would have a very different interface compared to the other fields such as subnets.

I am ok with whatever you guys suggest. If you want `any` for both levels then I can make the change.

Reply:
On further thought, I think if users want to ignore type checking they can just wrap the fw submodule like we do with main module. On a side note, I don't think defaults are slated till `0.14.0` 😞